### ISSUE TITLE
RAFT replication: ER_READONLY: Can't modify data on a read-only instance - state is election follower with term 2

### ISSUE THREAD

**Issue by cdump** [tarantool/tarantool#8589]
**Bug description**

Cluster of 3 nodes with RAFT enabled failes to bootstrap if one node starts after `TT_REPLICATION_CONNECT_TIMEOUT`

* OS: Linux
* OS Version: n/a
* Architecture: amd64
* Docker tag: tarantool/tarantool:2.10.6 (also affects 2.10.4 and probably other versions too)
```
Tarantool 2.10.6-0-g3990f976b
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: GNU-10.3.1
C_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/usr/src/tarantool=. -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type
CXX_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/usr/src/tarantool=. -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type
```

**Steps to reproduce**

_repr.lua:_
```lua
#!/usr/bin/env tarantool

box.cfg{}
pcall(function() box.schema.user.grant('guest', 'execute,read,write,create', 'universe') end)
```

_repr.sh_
```sh
#!/bin/bash

for I in `seq 1 3`; do
    docker run \
        -eTT_LISTEN="127.0.0.1:3200${I}"                                   \
        -eTT_ELECTION_TIMEOUT="0.5"                                        \
        -eTT_ELECTION_MODE="candidate"                                     \
        -eTT_REPLICATION_CONNECT_TIMEOUT=10                                \
        -eTT_REPLICATION_CONNECT_QUORUM=2                                  \
        -eTT_REPLICATION="127.0.0.1:32001,127.0.0.1:32002,127.0.0.1:32003" \
        --rm --net host -v $(pwd)/repr.lua:/repr.lua \
        --name tnt-${I} -d tarantool/tarantool:2.10.6 /repr.lua
    [[ ${I} == 1 ]] && sleep 7;
    [[ ${I} == 2 ]] && sleep 5;
done

sleep 5;

for I in `seq 1 3`; do
    echo "===== ${I} ====="
    docker exec -it tnt-${I} sh -c "echo 'box.info.ro, box.info.replication' | tarantoolctl connect 127.0.0.1:3200${I}"
    echo ""
    docker logs tnt-${I}
    echo "// -------------"
done;

for I in `seq 1 3`; do
    docker stop tnt-${I}
done
```

**Actual behavior**
Sometimes it's ok, but sometimes (1 out of 2..3 runs) it fails with the following logs.
If you remove the delay between node starts, everything becomes fine (or just happens with a lower probability).

**Full log:** [log.txt](https://github.com/tarantool/tarantool/files/11294877/log.txt)
```
===== 1 =====
Failed to connect to Tarantool

...
2023-04-21 10:24:00.673 [1] main/111/applier/127.0.0.1:32002 box.cc:238 E> ER_READONLY: Can't modify data on a read-only instance - state is election follower with term 2, leader is 1 (1131c209-6b1c-4577-a94d-431e337d35f5)
2023-04-21 10:24:00.673 [1] main/111/applier/127.0.0.1:32002 I> will retry every 1.00 second


...
===== 2 =====
...
2023-04-21 10:24:04.681 [1] main/122/main box.cc:238 E> ER_READONLY: Can't modify data on a read-only instance - state is election follower with term 2, leader is 1 (1131c209-6b1c-4577-a94d-431e337d35f5)

===== 3 =====
...
2023-04-21 10:24:00.915 [1] main/111/applier/127.0.0.1:32002 I> RAFT: fencing resumed
2023-04-21 10:24:00.915 [1] main/111/applier/127.0.0.1:32002 I> leaving orphan mode
```

**Expected behavior**
- Live 1 node (working `tarantoolctl connect`)
- No errors on 1st and 2nd nodes

**Comment by sergos** [comment]
The reason you see `ER_READONLY` is you run the grant command on a replica, not the master. Granting is a write into a system space and happens only if permissions are not granted yet. Subsequent tries to grant the same rights do not result into a write, so if it happens on the replica _after_ it happens on the master you see no error. 
Have a look into [this](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_once/) - perhaps it is something you expected?

**Comment by sergepetrenko** [comment]
Thanks for the report!

<details><summary>Here's what happens judging by the logs:</summary>

Instance 1 is started at 10:23:44.659, it has uuid dc8615bf-2a81-477a-b9d5-48246b10de5b.
Instance 2 is started at 10:23:53.110, it has uuid 85fca321-8b07-4ab2-ba1d-78c2794f10bb.
Instance 1 sees instance 2, and one second later the `replication_connect_timeout` passes:
```
2023-04-21 10:23:53.668 [1] main/111/applier/127.0.0.1:32002 I> remote master 85fca321-8b07-4ab2-ba1d-78c2794f10bb at 127.0.0.1:32002 running Tarantool 2.10.6
2023-04-21 10:23:54.661 [1] main/103/repr.lua C> failed to connect to 1 out of 3 replicas
```
Instance 1 chooses instance 2 as the bootstrap leader, but instance 2 still waits for connections (for it the replication_connect_timeout would pass only at ~10:24:03), hence the ER_LOADING error in instance1 log:
```
2023-04-21 10:23:54.662 [1] main/103/repr.lua I> bootstrapping replica from 85fca321-8b07-4ab2-ba1d-78c2794f10bb at 127.0.0.1:32002   2023-04-21 10:23:54.662 [1] main/111/applier/127.0.0.1:32002 I> can't read row   
2023-04-21 10:23:54.662 [1] main/111/applier/127.0.0.1:32002 box.cc:3228 E> ER_LOADING: Instance bootstrap hasn't finished yet
2023-04-21 10:23:54.662 [1] main/111/applier/127.0.0.1:32002 I> will retry every 1.00 second
```
Now instance 3 is started at 10:23:59.747, it has uuid 1131c209-6b1c-4577-a94d-431e337d35f5.
It sees everyone and decides to become the bootstrap leader.

While this happens, instance 2 starts seeing instance 3 and decides to bootstrap off it:
```
2023-04-21 10:24:00.116 [1] main/112/applier/127.0.0.1:32003 I> remote master 1131c209-6b1c-4577-a94d-431e337d35f5 at 127.0.0.1:32003 running Tarantool 2.10.6
2023-04-21 10:24:00.117 [1] main/103/repr.lua I> connected to 3 replicas         
2023-04-21 10:24:00.117 [1] main/103/repr.lua I> bootstrapping replica from 1131c209-6b1c-4577-a94d-431e337d35f5 at 127.0.0.1:32003
```

When elections are enabled, the bootstrap leader also becomes the election leader in the first ever term (term 2),
so instance 3 becomes the election leader.

Now instance 1 retries bootstrapping from instance 2 and continuously fails with ER_READONLY:
```
2023-04-21 10:24:00.673 [1] main/111/applier/127.0.0.1:32002 I> can't read row   
2023-04-21 10:24:00.673 [1] main/111/applier/127.0.0.1:32002 box.cc:238 E> ER_READONLY: Can't modify data on a read-only instance -   state is election follower with term 2, leader is 1 (1131c209-6b1c-4577-a94d-431e337d35f5)
2023-04-21 10:24:00.673 [1] main/111/applier/127.0.0.1:32002 I> will retry every 1.00 second
```
</details>

This particular error is fixed by commit 09c18907cdca3c347e5f0c248f3ce5e2679beda6. It wasn't pushed to 2.10 due to test incompatibility. With the fix in your example instance 1 would retry joining to instance 3 and succeed.

A general recomendation from me, regardless of the fix, is to make `replication_connect_timeout` higher. So that all the nodes start during the timeout. Even with the ER_READONLY fix another error is possible.

Imagine the same start up sequence with replication_connect_timeout = 10.
If instance 1 chose itself as the bootstrap leader (and not instance 2), and instance 3 again chose itself, you would end up with 2 separate clusters. One bootstrapped by instance 1, and another bootstrapped by instance 3.

For the bootstrap procedure to stay correct every node must see (be able to reach) the same set of remote peers. So they should all start up in time interval `[first node start; first node start + replication_connect_timeout]`.

If this rule is not satisfied, all sorts of erroneous results are possible.
